### PR TITLE
[4.2] Improve nonce generation and add the nonce to included files too as well as improve generation

### PR DESCRIPTION
**Pull request by zero-24**
Pull Request for Issue #37461

### Summary of Changes

Improve nonce generation and add the nonce to included files too as well as improve generation

#### First test
- enable the nonce generation as well as the CSP with report only
- check that the nonces are only added to inline style and inline scripts
- apply the patch
- note that the nonce is also added to JS / css files included locally

IMPORTANT: When you are checking the HTML with the nonce present in the header modern browsers will alter the sourcecode generated and remove the actuall nonce but only keep the nonce keyword. This is correct and shows that it worked correctly.

#### seccond test
- disable the nonce generation
- check that the nonces are added to inline JS and css
- apply the patch
- check that with that patch it is no longer added to inlinestyle and inlinejs with the nonce option disabled.

### Actual result BEFORE applying this Pull Request

- Nonces are always set even when its not used in the CSP
- Nonces are only set to the inline js and inline css

### Expected result AFTER applying this Pull Request

- Nonces are only rendered when also used within the CSP
- Nonces are also added to included JS files

### Documentation Changes Required

n.a.

cc @sisko1990

### CODE REVIEW

**Comment by gopitzohlsen** [comment]
With nonce enabled, I get nonce values in the style-src directive without configuring {nounce} placeholder inside the style-src directive.
With nonce disabled, I get no nonces inside the style-src directive.

**Comment by richard67** [comment]
> With nonce enabled, I get nonce values in the style-src directive without configuring {nounce} placeholder inside the style-src directive. With nonce disabled, I get no nonces inside the style-src directive.

@gopitzohlsen Does that mean you have successfully tested this pull request? If so, could you go to the issue tracker here https://issues.joomla.org/tracker/joomla-cms/37942 and use the blue "Test this" button at the top left corner? Just use the button, then select your test result and then submit. Thanks in advance.

**Comment by gopitzohlsen** [comment]
I have tested this item :white_check_mark: successfully on 01e315ae4c306cc62ce07e807596f1b3065aaaeb<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37942">issues.joomla.org/tracker/joomla-cms/37942</a>.</sub>

**Comment by viocassel** [comment]
I have tested this item :white_check_mark: successfully on 01e315ae4c306cc62ce07e807596f1b3065aaaeb<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37942">issues.joomla.org/tracker/joomla-cms/37942</a>.</sub>

File: plugins/system/httpheaders/httpheaders.php [hunk 1]
```diff
@@ -138,8 +138,15 @@ public function __construct(&$subject, $config)
 	{
 		parent::__construct($subject, $config);
 
-		// Nonce generation
-		$this->cspNonce = base64_encode(bin2hex(random_bytes(64)));
+		$nonceEnabled = (int) $this->params->get('nonce_enabled', 0);
+
+		// Nonce generation when its enabled
```
Comment (Quy) [inline plugins/system/httpheaders/httpheaders.php]:
```suggestion
		// Nonce generation when it's enabled
```

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37942">issues.joomla.org/tracker/joomla-cms/37942</a>.</sub>

**Comment by bembelimen** [comment]
Same here, will move to 4.2. Thx

**Comment by roland-d** [comment]
Thanks everybody

**Pull request merged**